VideoMediaSampleRenderer drops more frames than it should.

jyavenard opened the pull request:
#### 88039a970adb50d9cf382ea84903911eef45ce3c
<pre>
VideoMediaSampleRenderer drops more frames than it should.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283626">https://bugs.webkit.org/show_bug.cgi?id=283626</a>
<a href="https://rdar.apple.com/140515014">rdar://140515014</a>

Reviewed by Eric Carlson.

Whenever we added a new video sample to the decoded queue we would call purgeDecodedSampleQueue()
which would re-schedule a run of purgeDecodedSampleQueue at the presentation end time of the same
at the top of the queue.
If we already had a pending schedule to display the next frame, that next frame would now never be displayed
and would be dropped on the next purgeDecodedSampleQueue() call.

We now only reschedule if we added a frame earlier than the next schedule.
Manually tested.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay): Record the time of the sample being displayed, and not the currentTime.
Otherwise it could lead to false negative and a frame could be enqueued more than once.
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurge): Added method, we only call purgeDecodedSampleQueue if no
schedule has been set, or if the video sample just added is earlier than the current set timer.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): When calling purgeDecodedSampleQueue, ensure than
it doesn&apos;t reschedule a call which could prevent displaying the next frame.

Canonical link: <a href="https://commits.webkit.org/287074@main">https://commits.webkit.org/287074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ff5773e40b356de8ec6acf6a8cdf23d7063c957

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78145 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82804 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29407 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80278 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5473 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61196 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19113 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51183 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/68012 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41510 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24756 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27745 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69638 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25112 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84167 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5511 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3727 "Found 1 new test failure: ipc/create-media-source-with-invalid-constraints-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69419 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5668 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67103 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68673 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17145 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12671 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10922 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5460 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/8212 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5449 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8881 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7238 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -520,9 +523,31 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
     return samplesPurged;
 }
 
+void VideoMediaSampleRenderer::maybeReschedulePurge()
```
This is more a `maybePurge` than anything else.

jyavenard replied:
yes, but in PR #37128 it really is just a new scheduling. So I will keep this name.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -635,7 +660,7 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
             return;
 
-        purgeDecodedSampleQueue();
+        purgeDecodedSampleQueue(ShouldReschedule::No);
```
Why do we need to purge it?
Is it an optimisation in case we have a frame with a duration, and there will be no other frame after that time.
And we want to merge the frame when the duration is passed?

jyavenard replied:
I don't actually believe we need it, in theory the head of the queue should always be the right frame for display.

This is the original code, and I did consider this potential issue:

The timerSource is always a bit late at the time its task run, When I added logging I noticed that the timebase's current time is always about 1 to 2ms late. Consistently.

Let's assume we have frame A starting at 0.000 and frame B starting at 0.017
So when the timer kicks in to purge frame A and display frame B, will actually occurs at around 0.01715 1.5ms late.
It is possible that the call to `copyDisplayedPixelBuffer` is made at 0.017 exactly, before the purge task run, we would now incorrectly retrieve frame A when B would have been expected.

Now it's probably not an issue for the majority of the case in reality as the player will not receive a notification that there's a new frame until after B frame is shown.
But the content process can call `copyDisplayedPixelBuffer` at any time.

And this is the reason why I left that call there, even when my first thought was just like yours.
> Why do we need to purge it?
Is it an optimisation in case we have a frame with a duration, and there will be no other frame after that time.
And we want to merge the frame when the duration is passed?

It's not. there's no such merge. `copyDisplayedPixelBuffer` will not retrieve the frame only if the start time of the frame is after the current time.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -520,9 +523,31 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
+
+    CMTime duration = PAL::CMSampleBufferGetOutputDuration(nextSample.get());
+    CMTime presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp(nextSample.get());
+    CMTime presentationEndTime = PAL::CMTimeAdd(presentationTime, duration);
```
I would be tempted to add a `shouldPergeSample()` that would do this check.
Then, we would write it:
```
if (!shouldHeadSample())
    return;
purgeDecodedSampleQueue(ShouldReschedule::Yes);
```

jyavenard committed:
VideoMediaSampleRenderer drops more frames than it should.
https://bugs.webkit.org/show_bug.cgi?id=283626
rdar://140515014

Reviewed by Eric Carlson.

Whenever we added a new video sample to the decoded queue we would call purgeDecodedSampleQueue()
which would re-schedule a run of purgeDecodedSampleQueue at the presentation end time of the same
at the top of the queue.
If we already had a pending schedule to display the next frame, that next frame would now never be displayed
and would be dropped on the next purgeDecodedSampleQueue() call.

We now only reschedule if we added a frame earlier than the next schedule.
Manually tested.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::maybeQueueFrameForDisplay): Record the time of the sample being displayed, and not the currentTime.
Otherwise it could lead to false negative and a frame could be enqueued more than once.
(WebCore::VideoMediaSampleRenderer::flushDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::maybeReschedulePurge): Added method, we only call purgeDecodedSampleQueue if no
schedule has been set, or if the video sample just added is earlier than the current set timer.
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer): When calling purgeDecodedSampleQueue, ensure than
it doesn't reschedule a call which could prevent displaying the next frame.

Canonical link: https://commits.webkit.org/287074@main
